### ISSUE TITLE
Loss reported by fit vs custom_loop

### ISSUE THREAD

**Issue by Flamefire** [tensorflow/tensorflow#45923]
## URL(s) with the issue:

https://www.tensorflow.org/tutorials/distribute/custom_training

## Description of issue (what needs changing):

I wanted to compare the `fit` performance vs a "custom loop" with a `MultiWorker` strategy according to the tutorials at https://www.tensorflow.org/tutorials/distribute/keras and https://www.tensorflow.org/tutorials/distribute/custom_training 

However I'm seeing differences in the loss reported. Examples for an MNIST dataset with a Lenet:
- fit: Eval loss: 32.09, Eval Accuracy: 0.9307
- custom loop: Eval loss: 1.742, Eval Accuracy: 0.965

Also the loss reported in the `logs` member of the callbacks seems to be higher by some factor compared to the custom_loop.

### Clear description

At https://www.tensorflow.org/tutorials/distribute/custom_training#define_the_loss_function I found the following:

> - Using tf.reduce_mean is not recommended. Doing so divides the loss by actual per replica batch size which may vary step to step.
> - This reduction and scaling is done automatically in keras `model.compile` and `model.fit`
> - ... SUM_OVER_BATCH_SIZE is disallowed because currently it would only divide by per replica batch size, and leave the dividing by number of replicas to the user, which might be easy to miss. So instead we ask the user do the reduction themselves explicitly.

So question here which isn't clear to me from reading the docs/tutorials: How is the loss calculated by `model.compile` and `model.fit` and is that expected to be the same as the one shown in the tutorial for custom_loops? 

What I recon is that it should be the average loss per sample over the whole batch (on all replicas). So the sum of all `per_example_loss` divided by the number of examples.
The custom_loop tutorial divides by the global batch size, i.e. the local batch size times the number of replicas. This would be wrong for the mentioned case where the "per replica batch size [...] may vary step to step." So this sounds like `tf.reduce_mean` would be useful with an additional step of averaging the `per_replica_losses` instead of summing them after the `strategy.run` part. Why is that not recommended?

**Comment by ravikyram** [comment]
@Flamefire 

Can you please share colab link or simple standalone code to reproduce the issue in our environment.It helps us in localizing the issue faster. Thanks!

**Comment by rmothukuru** [comment]
@Flamefire,
Can you please respond to the [above comment](https://github.com/tensorflow/tensorflow/issues/45923#issuecomment-749998924)? Thanks!

**Comment by Flamefire** [comment]
I created a minified example which trains the same model 2 times: once with the `fit` function and once with the custom loop. Both variants follow very closely the tutorial code and should train the same network on the same dataset for the same number of batches resulting in the same losses.

However the loss for the custom loop at the end is significantly lower than that from `fit`, see https://colab.research.google.com/drive/1G_r2pU9usKgB5U_MAEkBZ0eJyK48YMta?usp=sharing

**Comment by Flamefire** [comment]
Ok I found the issue(s):

- I was missing a `strategy.experimental_distribute_dataset`. Not sure if this is new as I started writing the code for TF 2.0
- `fit` reports the loss via the `Mean` metric, so the loss is (likely) always higher then what I manually had which reported the loss for the current batch(es)

It might be worth mentioning that in the tutorial for custom loops

**Comment by nikitamaia** [comment]
Hi @Flamefire, `strategy.experimental_distribute_dataset` should be mentioned everywhere custom training loops are dicussed. However, if you find somewhere it's missing let me know.

As for updating the custom loops tutorial, can you elaborate on what you think should be added and where?

**Comment by Flamefire** [comment]
> As for updating the custom training loops tutorial, can you elaborate on what you think should be added and where?

The missing link for me was that `model.fit` reports the loss as the Mean, also in callbacks. https://www.tensorflow.org/tutorials/distribute/custom_training#training_loop has `train_loss = total_loss / num_batches` so I guess that is similar enough, but seems like I missed it. Maybe using the `Mean` metric here would be a good thing to show so users can output the current mean loss inbetween batches when starting with the tutorial code?

Anyway at the current state nothing seems to be really missing, so I'd also be fine to close this now. Thanks a lot!

**Comment by nikitamaia** [comment]
Thanks! I will close this issue now.

**Comment by rohanmuplara** [comment]
@nikitamaia @Flamefire @rmothukuru 


I had some questions about distribute strategy custom training. https://www.tensorflow.org/tutorials/distribute/custom_training. 

1. "If you are using regularization losses in your model then you need to scale the loss value by number of replicas. You can do this by using the tf.nn.scale_regularization_loss function."

**Why are regularization loss treated different compared to other type of losses?**

2. "Using tf.reduce_mean is not recommended. Doing so divides the loss by actual per replica batch size which may vary step to step."

**How does the batch size vary step to step per replica? Also, I am not exactly sure why we would not tf.reduce mean if the batch size changes step instead of diving by the global batch size**
